cpp/src/groupby/sort/group_dense_rank_scan.cu

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/logical.h>
+
+namespace cudf {
+namespace groupby {
+namespace detail {
+namespace {
+template <bool nested_nulls>
+void generate_dense_rank_struct_comparisons(column_view const& order_by,
+                                            mutable_column_view out,
+                                            device_span<size_type const> group_labels,
+                                            cudf::device_span<size_type const> group_offsets,
+                                            rmm::cuda_stream_view stream)
+{
+  thrust::tabulate(
+    rmm::exec_policy(stream),
+    out.begin<size_type>(),
+    out.end<size_type>(),
+    [d_order_by = *column_device_view::create(order_by, stream),
```

Comment:
You should create the pointer outside of the lambda, otherwise it will be destroyed immediately and you may have a dangling pointer. So:
```
auto const order_by_ptr = column_device_view::create(order_by, stream);
...
thrust::tablulate(.....
[d_order_by = *order_by_ptr]....